File: python/twoSum.py
# Time -> 0(N) Space -> 0(N)
def twoSum(nums: list[int], target: int) -> list[int]:
    seen = {}
    for i in range(len(nums)):
        seen[nums[i]] = i

    for i in range(len(nums)):
        diff = target - nums[i]
        if diff in seen and i != seen[diff]:
            return [i, seen[diff]]

    return []


print(twoSum([2, 7, 11, 15], 9))

File: go/TwoSum/TwoSum.go
package main

import "fmt"

// Time -> 0(N) Space -> 0(N)
func twoSums(nums []int, target int) []int {
	set := make(map[int]int, 0)
	for i, val := range nums {
		set[val] = i
	}

	for i, val := range nums {
		diff := target - val
		if val, ok := set[diff]; ok && i != val {
			return []int{val, i}
		}
	}
	return []int{}
}

func main() {
	fmt.Println(twoSums([]int{2, 7, 11, 15}, 9))
}
